### PR TITLE
feat: interim supervision order (FPLA-983)

### PR DESCRIPTION
**Pull request by joez-f**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-983

and test https://tools.hmcts.net/jira/browse/FPLA-981 here too

### Change description ###

Added subtype choice to order type page so users can select what type of supervision order they want (final/interim). This changes the document and filename accordingly, etc.

Test #799 and #803 together in this PR

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java [hunk 1]
```diff
@@ -96,11 +96,13 @@ public OrderTypeAndDocument buildOrderTypeAndDocument(OrderTypeAndDocument typeA
                 break;
             case SUPERVISION_ORDER:
                 orderBuilder.title(null);
-                ofNullable(orderMonths)
-                    .map(i -> time.now().plusMonths(orderMonths))
-                    .map(dateTime -> dateFormatterService.formatLocalDateTimeBaseUsingFormat(
-                        dateTime, "h:mma, d MMMM y"))
-                    .ifPresent(orderBuilder::expiryDate);
+                if (typeAndDocument.getSubtype() == FINAL) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: Is the if required?
If the user selects interim then orderMonths would be null, as such the expiryDate wouldn't be added.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java [hunk 2]
```diff
@@ -145,11 +147,21 @@ public OrderTypeAndDocument buildOrderTypeAndDocument(OrderTypeAndDocument typeA
...
-                    .put("orderDetails", getFormattedFinalSupervisionOrderDetails(getChildrenDetails(caseData).size(),
-                        caseData.getCaseLocalAuthority(), caseData.getOrderMonths()));
+                if (subtype == INTERIM) {
+                    orderTemplateBuilder
+                        .put("orderTitle", orderTypeAndDocument.getFullType(INTERIM))
+                        .put("childrenAct", "Section 38 and Paragraphs 1 and 2 Schedule 3 Children Act 1989")
+                        .put("orderDetails",
+                            getFormattedInterimSupervisionOrderDetails(getChildrenDetails(caseData).size(),
+                                caseData.getCaseLocalAuthority()));
+                } else if (subtype == FINAL) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: Minor but could just be else, it is either going to be `INTERIM` or `FINAL` and nothing else

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java (hunk 1 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: oh right, yea then not needed 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java [hunk 3]
```diff
@@ -206,6 +218,13 @@ private String getFormattedCareOrderDetails(int numOfChildren,
             isInterim ? " until the end of the proceedings." : ".");
     }
 
+    private String getFormattedInterimSupervisionOrderDetails(int numOfChildren, String caseLocalAuthority) {
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: Again there is room to test this in isolation if you feel it's needed.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java [hunk 4]
```diff
@@ -300,18 +305,28 @@ void shouldRemoveOrderPropertiesWhenTheyExistInCaseDetails() {
...
-                                + "date of this order until %s.", formattedDateTime));
+                    .put("orderType", SUPERVISION_ORDER);
+                if (subtype == INTERIM) {
+                    expectedMap
+                        .put("orderTitle", "Interim supervision order")
+                        .put("childrenAct", "Section 38 and Paragraphs 1 and 2 Schedule 3 Children Act 1989")
+                        .put("orderDetails",
+                            "It is ordered that Example Local Authority supervises the child until the end of the "
+                                + "proceedings");
+                } else if (subtype == FINAL) {
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java]: Same comment about `if` here

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java [hunk 5]
```diff
@@ -300,18 +305,28 @@ void shouldRemoveOrderPropertiesWhenTheyExistInCaseDetails() {
...
+                    final String formattedDateTime =
+                        dateFormatterService.formatLocalDateTimeBaseUsingFormat(date.plusMonths(5),
+                            "h:mma 'on the' d'" + suffix + "' MMMM y");
+                    expectedMap
+                        .put("orderTitle", "Supervision order")
+                        .put("childrenAct", "Section 31 and Paragraphs 1 and 2 Schedule 3 Children Act 1989")
+                        .put("orderDetails",
+                            String.format(
+                                "It is ordered that Example Local Authority supervises the child for 5 months from the "
+                                    + "date of this order until %s.", formattedDateTime));
```
Comment (Jamesnelson117) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java]: This is no longer as clean as it once was, maybe extracting the logic within the if conditions may help?

Comment (joez-f) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java]: as per lewis comment I will remove the `else if (type == FINAL)`, but don't think it's worth extracting the logic

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java (hunk 3 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: Isn't it tested as part of testing the `getOrderTemplateData` method?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java [hunk 6]
```diff
@@ -145,11 +147,21 @@ public OrderTypeAndDocument buildOrderTypeAndDocument(OrderTypeAndDocument typeA
...
-                    .put("orderTitle", orderTypeAndDocument.getFullType())
-                    .put("childrenAct", "Section 31 and Paragraphs 1 and 2 Schedule 3 Children Act 1989")
-                    .put("orderDetails", getFormattedFinalSupervisionOrderDetails(getChildrenDetails(caseData).size(),
-                        caseData.getCaseLocalAuthority(), caseData.getOrderMonths()));
+                if (subtype == INTERIM) {
+                    orderTemplateBuilder
+                        .put("orderTitle", orderTypeAndDocument.getFullType(INTERIM))
+                        .put("childrenAct", "Section 38 and Paragraphs 1 and 2 Schedule 3 Children Act 1989")
+                        .put("orderDetails",
+                            getFormattedInterimSupervisionOrderDetails(getChildrenDetails(caseData).size(),
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: We seem to call `getChildrenDetails(caseData)` 4 times now, probably best to extract to a variable

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java (hunk 5 above)
Comment (joez-f) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java]: Not removing the `else if` anymore as better to be safer, just on the off chance the enum is extended and we add more types. Still dont think its worth extracting the condition logic

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java (hunk 6 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderService.java]: extracted to variable

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java (hunk 4 above)
Comment (joez-f) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/GeneratedOrderServiceTest.java]: We decided not to change it, better to be safe

**Pull request merged**